### PR TITLE
Fix for handling the dtype conversion of object to string safely

### PR DESCRIPTION
**Pull request by dimosped**
Fixes issue: #157 
We have breakages due to this issue, this is a high priority fix.

### CODE REVIEW

**Comment by bmoscon** [comment]
@dimosped looks good

File: arctic/serialization/numpy_records.py [hunk 1]
```diff
@@ -29,12 +29,15 @@ def set_fast_check_df_serializable(config):
 
 def _to_primitive(arr, string_max_len=None):
     if arr.dtype.hasobject:
-        if len(arr) > 0:
-            if isinstance(arr[0], Timestamp):
+        if len(arr) > 0 and isinstance(arr[0], Timestamp):
                 return np.array([t.value for t in arr], dtype=DTN64_DTYPE)
         if string_max_len:
```
Comment (yschimke) [inline arctic/serialization/numpy_records.py]: nit: either add a blank line or make this an elif.

File: tests/unit/serialization/test_numpy_records.py [hunk 2]
```diff
@@ -100,3 +101,39 @@ def test_can_convert_to_records_without_objects_returns_true_otherwise(fast_seri
             store._to_records.assert_called_once_with(sentinel.df)
     finally:
         anr._FAST_CHECK_DF_SERIALIZABLE = orig_config
+
+
+@pytest.mark.parametrize("fast_serializable_check", (False, True))
+def test_can_convert_to_records_mixed_object_column_string_nan(fast_serializable_check):
+    orig_config = anr._FAST_CHECK_DF_SERIALIZABLE
```
Comment (yschimke) [inline tests/unit/serialization/test_numpy_records.py]: Is it possible to move all this boilerplate up into the annotation processing, suite generation?  Avoid copying and pasting?

File: arctic/serialization/numpy_records.py [hunk 3]
```diff
@@ -29,12 +29,15 @@ def set_fast_check_df_serializable(config):
...
-            return np.array(arr.astype('U{:d}'.format(string_max_len)))
-        return np.array(list(arr))
+            str_array = np.array(arr.astype('U{:d}'.format(string_max_len)))
```
Comment (yschimke) [inline arctic/serialization/numpy_records.py]: Is there a test case confirming this branch is handled?

File: CHANGES.md [hunk 4]
```diff
@@ -1,13 +1,15 @@
 ## Changelog
 
 ### 1.70
+  * Bugfix: #157 Preserve NaN and don't blindly stringify columns that don't only contain strings
```
Comment (lwfitzgerald) [inline CHANGES.md]: There's a lot of "don't"s in here making it a bit difficult to read.

Maybe this instead?
```
* Bugfix: #157 Preserve NaNs in columns containing strings
```

Comment (dimosped) [reply CHANGES.md]: yeah, language is not great, but this fix is more generic than NaNs, will change it to smth like:

Bugfix: #157 Assure that serialized dataframes remain value-equivalent (e.g. preserve NaNs in mixed string columns)

File: arctic/serialization/numpy_records.py (hunk 1 above)
Comment (dimosped) [reply arctic/serialization/numpy_records.py]: changed

File: tests/unit/serialization/test_numpy_records.py (hunk 2 above)
Comment (dimosped) [reply tests/unit/serialization/test_numpy_records.py]: refactored, looks cleaner now, thanks

File: arctic/serialization/numpy_records.py (hunk 3 above)
Comment (dimosped) [reply arctic/serialization/numpy_records.py]: done

File: CHANGES.md (hunk 4 above)
Comment (dimosped) [reply CHANGES.md]: don

File: arctic/serialization/numpy_records.py (hunk 1 above)
Comment (lwfitzgerald) [reply arctic/serialization/numpy_records.py]: Still indented by an extra level

Comment (dimosped) [reply arctic/serialization/numpy_records.py]: thanks @lwfitzgerald  , fixed now

**Pull request merged**